### PR TITLE
Metrics: add MetricsProducer and convert stats

### LINKED ISSUE #335: Add Metrics APIs as per "metrics.proto"
Add Metrics package and below classes based on [metrics.proto](https://github.com/census-instrumentation/opencensus-proto/blob/master/src/opencensus/proto/metrics/v1/metrics.proto). Although Metrics proto is still being finalized, We think it is safe to start working on value classes like LabelKey, TimeSeries, Point etc.

- [x] Metrics: Add LabelKey and LabelValue.
- [x] Metrics: Add Value and Point.
- [x] Metrics: Add TimeSeries.
- [x] Metrics: Add MetricDescriptor.
- [x] Metrics: Add Metric.
- [x] Metrics: Add Summary and Distribution.
- [x] Metrics: Produce and store Metrics in Stats impl.

Counterpart classes available in Java for reference : https://github.com/census-instrumentation/opencensus-java/tree/master/api/src/main/java/io/opencensus/metrics

### PR DESCRIPTION
**Pull request by c24t**
This PR adds a `MetricProducer` class and adds the plumbing to allow `Stats` to produce metrics, building on #469. It also fixes a bug converting `DistributionAggregationData`s without bucket bounds into metrics `Points`.

Closes #335.

### CODE REVIEW

File: opencensus/stats/stats.py [hunk 1]
```diff
@@ -12,24 +12,29 @@
...
     """Stats defines a View Manager and a Stats Recorder in order for the
     collection of Stats
     """
+
     def __init__(self):
-        self._stats_recorder = StatsRecorder()
-        self._view_manager = ViewManager()
-
-    @property
-    def stats_recorder(self):
```
Comment (c24t) [inline opencensus/stats/stats.py]: I'd understand making an attribute protected and then exposing it as a property if we were returning a copy, but we don't seem to be doing that here.

Comment (c24t) [reply opencensus/stats/stats.py]: https://github.com/census-instrumentation/opencensus-python/issues/477

File: opencensus/stats/stats.py [hunk 2]
```diff
@@ -12,24 +12,29 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from datetime import datetime
+
+from opencensus.metrics.export.metric_producer import MetricProducer
```
Comment (c24t) [inline opencensus/stats/stats.py]: I kept this consistent with other imports, but we should decide on class- or module-level imports across the library.

Comment (c24t) [reply opencensus/stats/stats.py]: #478

File: opencensus/metrics/export/metric_producer.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        :type metric_producer: :class: 'MetricProducer'
+        :param metric_producer: The metric producer to remove.
+        """
+        if metric_producer is None:
+            raise ValueError
+        try:
+            with self.mp_lock:
+                self.metric_producers.remove(metric_producer)
+        except KeyError:
+            pass
```
Comment (mayurkale22) [inline opencensus/metrics/export/metric_producer.py]: ```getAllMetricProducer``` is missing?

File: opencensus/metrics/export/metric_producer.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        with self.mp_lock:
+            self.metric_producers.add(metric_producer)
+
+    def remove(self, metric_producer):
+        """Remove a metric producer.
+
+        :type metric_producer: :class: 'MetricProducer'
+        :param metric_producer: The metric producer to remove.
+        """
+        if metric_producer is None:
```
Comment (mayurkale22) [inline opencensus/metrics/export/metric_producer.py]: I am not sure If we want to do None check here...

File: opencensus/metrics/export/metric_producer.py (hunk 3 above)
Comment (mayurkale22) [reply opencensus/metrics/export/metric_producer.py]: I see, ok.

Comment (c24t) [reply opencensus/metrics/export/metric_producer.py]: Added in fdd002f, if we go to the trouble of adding a lock to the class we might as well use it here.

**Pull request merged**